File: app/src/main/java/emerald/apps/fairychess/model/board/Movement.java
package emerald.apps.fairychess.model.board;

import java.util.ArrayList;
import java.util.List;

public class Movement {
    private MovementNotation movementNotation;
    private int sourceFile;
    private int sourceRank;
    private int targetFile;
    private int targetRank;

    public Movement(MovementNotation movementNotation, int sourceFile, int sourceRank, int targetFile, int targetRank) {
        this.movementNotation = movementNotation;
        this.sourceFile = sourceFile;
        this.sourceRank = sourceRank;
        this.targetFile = targetFile;
        this.targetRank = targetRank;
    }

    public Movement(int sourceFile, int sourceRank, int targetFile, int targetRank) {
        this(new MovementNotation("", new ArrayList<>(), "", new ArrayList<>(), ""), sourceFile, sourceRank, targetFile, targetRank);
    }

    public Movement(MovementNotation movementNotation, Coordinate source, int targetFile, int targetRank) {
        this(movementNotation, source.getFile(), source.getRank(), targetFile, targetRank);
    }

    public Movement(Coordinate source, int targetFile, int targetRank) {
        this(source.getFile(), source.getRank(), targetFile, targetRank);
    }

    public Movement(Coordinate source, Coordinate target) {
        this(source.getFile(), source.getRank(), target.getFile(), target.getRank());
    }

    public Coordinate getSourceCoordinate() {
        return new Coordinate(sourceFile, sourceRank);
    }

    public Coordinate getTargetCoordinate() {
        return new Coordinate(targetFile, targetRank);
    }

    public int getRankDif() {
        return Math.abs(targetRank - sourceRank);
    }

    public int getFileDif() {
        return Math.abs(targetFile - sourceFile);
    }

    public int getSignRank() {
        return (int) Math.signum(targetRank - sourceRank);
    }

    public int getTargetFile(){
        return this.targetFile;
    }

    public int getTargetRank(){
        return this.targetRank;
    }

    public int getSourceFile(){
        return this.sourceFile;
    }

    public int getSourceRank(){
        return this.sourceRank;
    }


    @Override
    public boolean equals(Object other) {
        if (other instanceof Movement) {
            Movement movement = (Movement) other;
            return sourceFile == movement.sourceFile &&
                    sourceRank == movement.sourceRank &&
                    targetFile == movement.targetFile &&
                    targetRank == movement.targetRank &&
                    movementNotation.equals(movement.movementNotation);
        }
        return super.equals(other);
    }

    public static Movement emptyMovement() {
        return new Movement(0, 0, 0, 0);
    }

    public static String fromMovementToString(Movement movement) {
        return movement.sourceFile + "_" + movement.sourceRank + "_" + movement.targetFile + "_" + movement.targetRank;
    }

    public static Movement fromStringToMovement(String string) {
        String[] coordinates = string.split("_");
        if (coordinates.length == 4) {
            int sourceFile = Integer.parseInt(coordinates[0]);
            int sourceRank = Integer.parseInt(coordinates[1]);
            int targetFile = Integer.parseInt(coordinates[2]);
            int targetRank = Integer.parseInt(coordinates[3]);
            return new Movement(sourceFile, sourceRank, targetFile, targetRank);
        }
        return new Movement(-1, -1, -1, -1);
    }

    public static String fromMovementListToString(List<Movement> movements) {
        StringBuilder returnString = new StringBuilder();
        for (Movement movement : movements) {
            returnString.append(movement.sourceFile).append("_").append(movement.sourceRank)
                    .append("_").append(movement.targetFile).append("_").append(movement.targetRank);
            if (movement != movements.get(movements.size() - 1)) returnString.append(";");
        }
        return returnString.toString();
    }

    public static List<Movement> fromStringToMovementList(String string) {
        List<Movement> movementList = new ArrayList<>();
        for (String substring : string.split(";")) {
            if (substring.split("_").length == 4) {
                int sourceFile = Integer.parseInt(substring.split("_")[0]);
                int sourceRank = Integer.parseInt(substring.split("_")[1]);
                int targetFile = Integer.parseInt(substring.split("_")[2]);
                int targetRank = Integer.parseInt(substring.split("_")[3]);
                movementList.add(new Movement(sourceFile, sourceRank, targetFile, targetRank));
            }
        }
        return movementList;
    }

    public String asString(String playerColor) {
        return playerColor + ": " + sourceFile + "_" + sourceRank + "_" + targetFile + "_" + targetRank;
    }

    public String asString2(String playerColor) {
        return playerColor + ": " + getLetterFromInt(sourceFile) + sourceRank + "_" + getLetterFromInt(targetFile) + targetRank;
    }

    public String getLetterFromInt(int intValue) {
        switch (intValue) {
            case 0: return "A";
            case 1: return "B";
            case 2: return "C";
            case 3: return "D";
            case 4: return "E";
            case 5: return "F";
            case 6: return "G";
            case 7: return "H";
            default: return "";
        }
    }
}

File: app/src/main/java/emerald/apps/fairychess/model/board/Movement.kt
package emerald.apps.fairychess.model.board

import java.util.*
import kotlin.math.abs
import kotlin.math.sign

open class Movement(val movementNotation : MovementNotation = MovementNotation("", emptyList(),"",emptyList(),"")
                    , val sourceFile : Int
                    , val sourceRank : Int
                    , val targetFile : Int
                    , val targetRank : Int) {

    constructor(sourceFile : Int, sourceRank : Int, targetFile : Int, targetRank : Int)
            : this(MovementNotation("", emptyList(),"",emptyList(),""), sourceFile, sourceRank, targetFile, targetRank)

    constructor(movementNotation: MovementNotation, source: Coordinate, targetFile: Int, targetRank: Int)
            : this(movementNotation, source.file, source.rank, targetFile, targetRank)
    constructor(source: Coordinate, targetFile: Int, targetRank: Int)
            : this(source.file, source.rank, targetFile, targetRank)
    constructor(source: Coordinate, target: Coordinate)
            : this(source.file, source.rank, target.file, target.rank)

    fun getSourceCoordinate(): Coordinate {
        return Coordinate(sourceFile, sourceRank)
    }

    fun getTargetCoordinate(): Coordinate {
        return Coordinate(targetFile, targetRank)
    }

    fun getRankDif(): Int {
        return abs(targetRank - sourceRank)
    }

    fun getFileDif(): Int {
        return abs(targetFile - sourceFile)
    }

    fun getSignRank(): Int {
        return sign(targetRank.toDouble() - sourceRank.toDouble()).toInt()
    }

    override fun equals(other: Any?): Boolean {
        return if(other is Movement) {
            (sourceFile == other.sourceFile
                    && sourceRank == other.sourceRank
                    && targetFile == other.targetFile
                    && targetRank == other.targetRank
                    && movementNotation == other.movementNotation
                    )
        } else super.equals(other)
    }

    companion object {
        fun emptyMovement(): Movement {
            return Movement(sourceFile = 0, sourceRank = 0, targetFile = 0, targetRank = 0)
        }

        fun fromMovementToString(movement: Movement): String {
            return movement.sourceFile.toString() + "_" + movement.sourceRank + "_" + movement.targetFile.toString() + "_" + movement.targetRank
        }

        fun fromStringToMovement(string: String): Movement {
            val coordinates = string.split("_")
            if(coordinates.size == 4) {
                val sourceFile = coordinates[0].toInt()
                val sourceRank = coordinates[1].toInt()
                val targetFile = coordinates[2].toInt()
                val targetRank = coordinates[3].toInt()
                return Movement(sourceFile = sourceFile, sourceRank = sourceRank, targetFile = targetFile, targetRank = targetRank)
            }
            return Movement(sourceFile = -1, sourceRank = -1, targetFile = -1, targetRank = -1)
        }

        fun fromMovementListToString(movements: List<Movement>): String {
            val returnString = StringBuilder("")
            for(movement in movements) {
                returnString.append(
                    movement.sourceFile.toString() + "_" + movement.sourceRank
                            + "_" + movement.targetFile.toString() + "_" + movement.targetRank)
                if(movement != movements.last()) returnString.append(";")
            }
            return returnString.toString()
        }

        fun fromStringToMovementList(string: String): List<Movement> {
            val movementList = mutableListOf<Movement>()
            for(substring in string.split(";")) {
                if(substring.split("_").size == 4) {
                    val sourceFile = substring.split("_")[0].toInt()
                    val sourceRank = substring.split("_")[1].toInt()
                    val targetFile = substring.split("_")[2].toInt()
                    val targetRank = substring.split("_")[3].toInt()
                    movementList.add(Movement(sourceFile = sourceFile, sourceRank = sourceRank, targetFile = targetFile, targetRank = targetRank))
                }
            }
            return movementList
        }
    }

    fun asString(playerColor: String): String {
        return playerColor + ": " + sourceFile.toString() + "_" + sourceRank + "_" + targetFile + "_" + targetRank
    }

    fun asString2(playerColor: String): String {
        return playerColor + ": " + getLetterFromInt(sourceFile) + sourceRank + "_" + getLetterFromInt(targetFile) + targetRank
    }

    fun getLetterFromInt(int: Int): String {
        when(int) {
            0 -> return "A"
            1 -> return "B"
            2 -> return "C"
            3 -> return "D"
            4 -> return "E"
            5 -> return "F"
            6 -> return "G"
            7 -> return "H"
        }
        return ""
    }
}

class PromotionMovement(movementNotation: MovementNotation = MovementNotation("", emptyList(),"",emptyList(),"")
                        , sourceFile: Int
                        , sourceRank: Int
                        , targetFile: Int
                        , targetRank: Int
                        , var promotion: String) : Movement(movementNotation, sourceFile, sourceRank, targetFile, targetRank) {

    constructor(sourceFile: Int, sourceRank: Int, targetFile: Int, targetRank: Int, promotion: String)
            : this(MovementNotation("", emptyList(),"",emptyList(),""), sourceFile, sourceRank, targetFile, targetRank, promotion)

    companion object {
        fun fromMovementToString(promotionMovement: PromotionMovement): String {
            return promotionMovement.sourceFile.toString() + "_" + promotionMovement.sourceRank + "_" +
                    promotionMovement.targetFile.toString() + "_" + promotionMovement.targetRank + "_" + promotionMovement.promotion
        }

        fun fromStringToMovement(string: String): Movement {
            val coordinates = string.split("_")
            if (coordinates.size == 5) {
                val sourceFile = coordinates[0].toInt()
                val sourceRank = coordinates[1].toInt()
                val targetFile = coordinates[2].toInt()
                val targetRank = coordinates[3].toInt()
                val promotion = coordinates[4]
                return PromotionMovement(
                    sourceFile = sourceFile,
                    sourceRank = sourceRank,
                    targetFile = targetFile,
                    targetRank = targetRank,
                    promotion = promotion
                )
            }
            return Movement(sourceFile = -1, sourceRank = -1, targetFile = -1, targetRank = -1)
        }
    }
}

class MovementNotation(val grouping: String, val conditions: List<String>, val movetype: String, val distances: List<String>, val direction: String) {
    // MovementNotation class remains unchanged as it doesn't deal with coordinate ordering

    override fun equals(other: Any?): Boolean {
        return if(other is MovementNotation) {
            (grouping == other.grouping
                    && conditions == other.conditions
                    && movetype == other.movetype
                    && distances == other.distances
                    && direction == other.direction
                    )
        } else super.equals(other)
    }

    override fun toString(): String {
        return grouping + conditions.toString() + movetype + distances + direction
    }

    companion object {
        val KING = MovementNotation("", listOf(),"", listOf("1"),"*")
        val PAWN_ENPASSANTE = MovementNotation("", listOf(),"", listOf("1"),"EN_PASSANTE")
        val CASTLING_SHORT_WHITE = MovementNotation("", listOf(),"CASTLING_SHORT_WHITE", listOf(),"")
        val CASTLING_LONG_WHITE = MovementNotation("", listOf(),"CASTLING_LONG_WHITE", listOf(),"")
        val CASTLING_SHORT_BLACK = MovementNotation("", listOf(),"CASTLING_SHORT_BLACK", listOf(),"")
        val CASTLING_LONG_BLACK = MovementNotation("", listOf(),"CASTLING_LONG_BLACK", listOf(),"")

        // parseMovementString function remains unchanged as it doesn't deal with coordinate ordering
        fun parseMovementString(movementString: String): List<MovementNotation> {
            if(movementString.isEmpty()) return emptyList()
            val movementList = mutableListOf<MovementNotation>()
            val movementArray = movementString.split(",")
            for(submovement in movementArray) {
                var submovementString = submovement
                var grouping = ""
                var conditions = mutableListOf<String>()
                var movetype = ""
                var distances = mutableListOf<String>()
                var direction = ""
                //move type
                if(submovementString.contains("~")) {movetype = "~";submovementString = submovementString.replace("~","")}
                if(submovementString.contains("^")) {movetype = "^";submovementString = submovementString.replace("^","")}
                if(submovementString.contains("g")) {movetype = "g";submovementString = submovementString.replace("g","")}
                //grouping
                if(submovementString.contains("/")) {grouping = "/";submovementString = submovementString.replace("/","")}
                if(submovementString.contains("&")) {grouping = "&";submovementString = submovementString.replace("&","")}
                if(submovementString.contains(".")) {grouping = ".";submovementString = submovementString.replace(".","")}
                //move conditions
                if(submovementString.contains("i")) {conditions.add("i");submovementString = submovementString.replace("i","")}
                if(submovementString.contains("c")) {conditions.add("c");submovementString = submovementString.replace("c","")}
                if(submovementString.contains("o")) {conditions.add("o");submovementString = submovementString.replace("o","")}
                //direction
                if(submovementString.contains(">=")) {direction = ">=";submovementString = submovementString.replace(">=","")}
                if(submovementString.contains("<=")) {direction = "<=";submovementString = submovementString.replace("<=","")}
                if(submovementString.contains("<>")) {direction = "<>";submovementString = submovementString.replace("<>","")}
                if(submovementString.contains("=")) {direction = "=";submovementString = submovementString.replace("=","")}
                if(submovementString.contains("X>")) {direction = "X>";submovementString = submovementString.replace("X>","")}
                if(submovementString.contains("X<")) {direction = "X<";submovementString = submovementString.replace("X<","")}
                if(submovementString.contains("X")) {direction = "X";submovementString = submovementString.replace("X","")}
                if(submovementString.contains(">")) {direction = ">";submovementString = submovementString.replace(">","")}
                if(submovementString.contains("<")) {direction = "<";submovementString = submovementString.replace("<","")}
                if(submovementString.contains("+")) {direction = "+";submovementString = submovementString.replace("+","")}
                if(submovementString.contains("*")) {direction = "*";submovementString = submovementString.replace("*","")}
                //distance
                if(grouping == "") {
                    if(submovementString.contains("n")) distances.add("n")
                    if(submovementString.contains("[0-9]".toRegex())) distances.add(submovementString.replace("\\D+".toString(),""))
                } else {
                    distances = submovementString.split("").toMutableList()
                    distances.removeAll(Collections.singleton(""))
                }
                movementList.add(
                    MovementNotation(
                        grouping,
                        conditions,
                        movetype,
                        distances.toList(),
                        direction
                    )
                )
            }
            return movementList
        }
    }
}